libs/client/src/realtime.ts

```diff
@@ -0,0 +1,233 @@
+ * realtime endpoint.
+ */
+export interface RealtimeConnection<Input> {
+  send(input: Input): void;
+
+  close(): void;
+}
+
+type ResultWithRequestId = {
+  request_id: string;
+};
+
+/**
+ * Options for connecting to the realtime endpoint.
+ */
+export interface RealtimeConnectionHandler<Output> {
+  /**
+   * The connection key. This is used to reuse the same connection
+   * across multiple calls to `connect`. This is particularly useful in
+   * contexts where the connection is established as part of a component
+   * lifecycle (e.g. React) and the component is re-rendered multiple times.
+   */
+  connectionKey?: string;
+
+  /**
+   * If `true`, the connection will only be established on the client side.
+   * This is useful for frameworks that reuse code for both server-side
+   * rendering and client-side rendering (e.g. Next.js).
+   */
+  clientOnly?: boolean;
+
+  /**
+   * The throtle duration in milliseconds. This is used to throtle the
+   * calls to the `send` function. Realtime apps usually react to user
+   * input, which can be very frequesnt (e.g. fast typing or mouse/drag movements).
+   *
+   * The default value is `64` milliseconds.
+   */
+  throttleInterval?: number;
+
+  /**
+   * Callback function that is called when a result is received.
+   * @param result - The result of the request.
+   */
+  onResult(result: Output & ResultWithRequestId): void;
+
+  /**
+   * Callback function that is called when an error occurs.
+   * @param error - The error that occurred.
+   */
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  onError?(error: ApiError<any>): void;
+}
+
+export interface RealtimeClient {
+  /**
+   * Connect to the realtime endpoint. The default implementation uses
+   * WebSockets to connect to fal function endpoints that support WSS.
+   *
+   * @param app the app alias or identifier.
+   * @param handler the connection handler.
+   */
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  connect<Input = any, Output = any>(
+    app: string,
+    handler: RealtimeConnectionHandler<Output>
+  ): RealtimeConnection<Input>;
+}
+
+function builRealtimeUrl(app: string): string {
+  const { host } = getConfig();
+  return `wss://${app}.${host}/ws`;
+}
+
+/**
+ * Get a token to connect to the realtime endpoint.
+ */
+async function getToken(app: string): Promise<string> {
+  return await dispatchRequest<any, string>(
+    'POST',
+    `https://${getRestApiUrl()}/tokens`,
+    {
+      allowed_apps: [app],
+      token_expiration: 60,
+    }
+  );
+}
+
+/**
+ * See https://www.rfc-editor.org/rfc/rfc6455.html#section-7.4.1
+ */
+const WebSocketErrorCodes = {
+  NORMAL_CLOSURE: 1000,
+  GOING_AWAY: 1001,
+};
+
+const connections = new Map<string, WebSocket>();
+
+async function getConnection(app: string, key: string): Promise<WebSocket> {
+  const url = builRealtimeUrl(app);
+  // const token = await getToken(app);
```

Comment:
So if we integrate (or enable) this, would this mean we have to make a new REST request (to generate a JWT token) even though we might have an unexpired token before every connection? Since we regularly close the underlying websockets, this might be a bit inefficient without some sort of caching to make sure we squeeze out as much as possible from individual JWT tokens (e.g. getToken with a TTL'd cache, if it is easy to implement) should help with perf a lot (in a world where re-connects happen often).

Reply:
Yes, once the wip impl is done, it will reuse the token as long as it's not expired.